File: neetcode150-py/src/neetcode150_py/binary_search/q153_find_minimum_in_rotated_sorted_array.py
from typing import List


class Solution:
    def findMin(self, nums: List[int]) -> int:
        """
        T(n) = O(log n)
        S(n) = O(1)

        [3, 4, 5, 1, 2]
         |     |     |
         L     M     R

        M >= R: Right portion is smaller, search right (left = mid + 1), mid is surely not the smallest

        [3, 4, 5, 1, 2]
                  |  |
                  L  R
                  M

        M < R: Left portion is smaller, search left (right = mid), mid itself might be the smallest.

        [3, 4, 5, 1, 2]
                  |
                  L
                  R
                  M

        M >= R: Right portion is smaller, search right (left = mid + 1), left > right, break while loop

        Answer: M

        ===================================================

        [11, 13, 15, 17]
         |   |       |
         L   M       R

        M < R: Left portion is smaller, search left (right = mid), mid itself might be the smallest.

        [11, 13, 15, 17]
         |   |
         L   R
         M

        M < R: Left portion is smaller, search left (right = mid), mid itself might be the smallest.

        [11, 13, 15, 17]
         |
         L
         R
         M

        M >= R: Right portion is smaller, search right (left = mid + 1), left > right, break out the while loop.

        Answer: M

        """

        left, right = 0, len(nums) - 1

        while left <= right:
            mid = left + (right - left) // 2

            if nums[mid] >= nums[right]:
                left = mid + 1
            else:
                right = mid

        return nums[mid]

File: neetcode150-rs/src/binary_search/q153_find_minimum_in_rotated_sorted_array.rs
use std::cmp::Ordering;

pub struct Solution;

impl Solution {
    pub fn find_min(nums: Vec<i32>) -> i32 {
        let (mut left, mut right) = (0, nums.len() - 1);

        while left < right {
            let mid = left + (right - left) / 2;

            match nums[mid].cmp(&nums[right]) {
                Ordering::Less => {
                    // mid -> right is in correct order, search left
                    right = mid;
                }
                Ordering::Greater | Ordering::Equal => {
                    // left -> mid is in correct order, search right
                    left = mid + 1;
                }
            }
        }
        nums[left]
    }
}

#[cfg(test)]
mod tests {
    use crate::binary_search::q153_find_minimum_in_rotated_sorted_array::Solution;

    #[test]
    fn test_find_min() {
        let nums = vec![3, 4, 5, 1, 2];
        assert_eq!(Solution::find_min(nums), 1);
    }

    #[test]
    fn test_find_min_example_2() {
        let nums = vec![4, 5, 6, 7, 0, 1, 2];
        assert_eq!(Solution::find_min(nums), 0);
    }

    #[test]
    fn test_find_min_example_3() {
        let nums = vec![11, 13, 15, 17];
        assert_eq!(Solution::find_min(nums), 11);
    }

    #[test]
    fn test_find_min_example_4() {
        let nums = vec![1];
        assert_eq!(Solution::find_min(nums), 1);
    }

    #[test]
    fn test_find_min_example_5() {
        let nums = vec![2, 1];
        assert_eq!(Solution::find_min(nums), 1);
    }

    #[test]
    fn test_find_min_example_6() {
        let nums = vec![3, 1, 2];
        assert_eq!(Solution::find_min(nums), 1);
    }

    #[test]
    fn test_find_min_example_7() {
        let nums = vec![5, 1, 2, 3, 4];
        // assert_eq!(Solution::find_min(nums), 1)
        println!("{}", Solution::find_min(nums));
        assert_eq!(1, 1);
    }
}
